lib/hocon/impl/parseable.rb

```diff
@@ -191,4 +241,293 @@ def raw_parse_value_from_io(io, origin, final_options)
+  class ParseableNotFound < Hocon::Impl::Parseable
+    def initialize(what, message, options)
+      super()
+      @what = what
+      @message = message
+      post_construct(options)
+    end
+
+    def custom_reader
+      raise Hocon::ConfigError::ConfigBugOrBrokenError, @message
+    end
+
+    def create_origin
+      Hocon::Impl::SimpleConfigOrigin.new_simple(@what)
+    end
+  end
+
+  def self.new_not_found(what_not_found, message, options)
+    ParseableNotFound.new(what_not_found, message, options)
+  end
+
+  # NOTE: skipping `ParseableReader` until we know we need it (probably should
+  #  have done that with `ParseableNotFound`)
+
+  class ParseableString < Hocon::Impl::Parseable
+    def initialize(string, options)
+      super()
+      @input = string
+      post_construct(options)
+    end
+
+    def custom_reader
+      if Hocon::Impl::ConfigImpl.trace_loads_enabled
+        self.class.trace("Loading config from a String: #{@input}")
+      end
+      # we return self here, which will cause `open` to be called on us, so
+      # we can provide an implementation of that.
+      self
+    end
+
+    def open
+      if block_given?
+        StringIO.open(@input) do |f|
+          yield f
+        end
+      else
+        StringIO.open(@input)
+      end
+    end
+
+    def create_origin
+      Hocon::Impl::SimpleConfigOrigin.new_simple("String")
+    end
+
+    def to_s
+      "#{self.class.name} (#{@input})"
+    end
+  end
+
+  def self.new_string(string, options)
+    ParseableString.new(string, options)
+  end
+
+  # NOTE: Skipping `ParseableURL` for now as we probably won't support this right away
+
+  class ParseableFile < Hocon::Impl::Parseable
+    def initialize(input, options)
+      super()
+      @input = input
+      post_construct(options)
+    end
+
+    def custom_reader
+      if Hocon::Impl::ConfigImpl.trace_loads_enabled
+        self.class.trace("Loading config from a String: #{@input}")
+      end
+      # we return self here, which will cause `open` to be called on us, so
+      # we can provide an implementation of that.
+      self
+    end
+
+    def open
+      if block_given?
+        File.open(@input) do |f|
+          yield f
+        end
+      else
+        File.open(@input)
+      end
+    end
+
+    def guess_syntax
+      Hocon::Impl::Parseable.syntax_from_extension(File.basename(@input))
+    end
+
+    def relative_to(filename)
+      sibling = nil
+      if Pathname.new(filename).absolute?
+        sibling = File.new(filename)
+      else
+        # this may return nil
+        sibling = Parseable.relative_to(@input, filename)
+      end
+      if sibling.nil?
+        nil
+      elsif sibling.exists?
+        self.class.trace("#{sibling" exists, so loading it as a file"}")
+        Parseable.new_file(sibling, options.set_origin_description(nil))
+      else
+        self.class.trace("#{sibling} does not exist, so trying it as a resource")
+        super(filename)
```

Comment:
Is this correct?  The java version is `return super.relativeTo(filename);`

Hmm, is that what `super